_posts/2025-09-01-semantic-router.md

```diff
@@ -0,0 +1,104 @@
+
+Over the past year, **hybrid reasoning and automatic routing** have emerged as some of the most discussed topics in the large-model ecosystem.
+
+Take **GPT-5** as an example. Its most significant breakthrough is not simply the number of parameters, but the introduction of **automatic routing and thinking quotas**:
+
+* **Light queries → Lightweight models**: For example, "Why is the sky blue?" does not require an expensive inference model.
+* **Complex/High-value queries → Advanced models**: Tasks such as legal analysis or financial simulations are routed to models with Chain-of-Thought capabilities.
+
+The principle behind this is often described as **per-token unit economics**.
+
+Every token generated must deliver value rather than being treated as pure computational expense.
+
+For example:
+
+* Free-tier users receive answers from lightweight models, keeping costs under control.
+* When a query indicates commercial intent (e.g., booking flights or finding legal services), it is routed to high-compute models or agent services directly integrated into transaction flows.
+
+In these cases, companies like OpenAI can participate in the value chain by taking a commission on completed transactions — transforming free usage from a cost center into a monetizable entry point.
+
+Other companies are adopting similar strategies:
+
+* **Anthropic Claude 3.7/4**: Combines "fast thinking" and "slow thinking" with user-controlled toggles.
+* **Google Gemini 2.5**: Introduces a *thinking budget*, giving enterprises fine-grained control over inference costs.
+* **Alibaba Qwen3**: Explores instruction-based switching between reasoning and non-reasoning modes.
+* **DeepSeek v3.1**: Implements a "single-model dual-mode" design, merging dialogue and reasoning.
+
+In short: the industry is entering an era where **no token should be wasted**.
+
+## Recent Research: vLLM Semantic Router
+
+Amid this shift toward hybrid reasoning, we focus on the **open-source inference engine vLLM**.
+
+While vLLM has become the de facto standard for deploying large models, it lacks fine-grained, semantic-level control — the ability to make routing decisions based on meaning rather than query type alone. Developers are often forced to either enable full inference (wasting computation) or disable it entirely (sacrificing accuracy).
+
+To address this, we propose the **vLLM Semantic Router**, which brings GPT-5-style "smart routing" to the open-source ecosystem.
+
+![](/assets/figures/semantic-router/architecture.png)
+
+### Architecture Design
+
+1. **Semantic Classification**: Uses a **ModernBERT** fine-tuned intent classifier to determine whether a query requires inference.
+2. **Smart Routing**:
+
+   * Simple queries → Fast inference mode.
+   * Complex queries → Chain-of-Thought for accurate reasoning.
+3. **High-Performance Engine**: Built with Rust and the Hugging Face Candle framework, enabling high concurrency and zero-copy efficiency.
```

Comment:
Implemented in **Rust** using the **Hugging Face Candle** framework, the engine achieves **high concurrency** and **zero-copy efficiency**, making it well-suited for large-scale serving.